sdk/storage/src/clients/storage_account_client.rs

```diff
@@ -0,0 +1,427 @@
+                add_if_exists(h, RANGE),
+                canonicalize_header(h),
+                canonicalized_resource(account, u)
+            )
+        }
+    }
+
+    // expected
+    // GET\n /*HTTP Verb*/
+    // \n    /*Content-Encoding*/
+    // \n    /*Content-Language*/
+    // \n    /*Content-Length (include value when zero)*/
+    // \n    /*Content-MD5*/
+    // \n    /*Content-Type*/
+    // \n    /*Date*/
+    // \n    /*If-Modified-Since */
+    // \n    /*If-Match*/
+    // \n    /*If-None-Match*/
+    // \n    /*If-Unmodified-Since*/
+    // \n    /*Range*/
+    // x-ms-date:Sun, 11 Oct 2009 21:49:13 GMT\nx-ms-version:2009-09-19\n
+    //                                  /*CanonicalizedHeaders*/
+    // /myaccount /mycontainer\ncomp:metadata\nrestype:container\ntimeout:20
+    //                                  /*CanonicalizedResource*/
+    //
+    //
+}
+
+fn canonicalize_header(h: &HeaderMap) -> String {
+    let mut v_headers = h
+        .iter()
+        .filter(|(k, _v)| k.as_str().starts_with("x-ms"))
+        .map(|(k, _)| k.as_str())
+        .collect::<Vec<_>>();
+    v_headers.sort_unstable();
+
+    let mut can = String::new();
+
+    for header_name in v_headers {
+        let s = h.get(header_name).unwrap().to_str().unwrap();
+        can = can + header_name + ":" + s + "\n";
+    }
+    can
+}
+
+// For table
+fn canonicalized_resource_table(account: &str, u: &url::Url) -> String {
+    format!("/{}{}", account, u.path())
+}
+
+fn canonicalized_resource(account: &str, u: &url::Url) -> String {
+    let mut can_res: String = String::new();
+    can_res += "/";
+    can_res += &account;
+
+    let paths = u.path_segments().unwrap();
+
+    {
+        let mut path = String::new();
+        for p in paths {
+            path.push('/');
+            path.push_str(&*p);
+        }
+
+        can_res += &path;
+    }
+    can_res += "\n";
+
+    // query parameters
+    let query_pairs = u.query_pairs(); //.into_owned();
+    {
+        let mut qps = Vec::new();
+        {
+            for qp in query_pairs {
+                trace!("adding to qps {:?}", qp);
+
+                // add only once
+                if !(qps.iter().any(|x: &String| x == &qp.0)) {
+                    qps.push(qp.0.into_owned());
+                }
+            }
+        }
+
+        qps.sort();
+
+        for qparam in qps {
+            // find correct parameter
+            let ret = lexy_sort(&query_pairs, &qparam);
+
+            // debug!("adding to can_res {:?}", ret);
+
+            can_res = can_res + &qparam.to_lowercase() + ":";
+
+            for (i, item) in ret.iter().enumerate() {
+                if i > 0 {
+                    can_res += ","
+                }
+                can_res += item;
+            }
+
+            can_res += "\n";
+        }
+    };
+
+    can_res[0..can_res.len() - 1].to_owned()
+}
+
+fn lexy_sort(vec: &url::form_urlencoded::Parse, query_param: &str) -> Vec<String> {
```

Comment:
You can probably easily just return `Vec<&str>` instead to avoid creating new strings.

Reply:
Good catch! I had to return a `Cow<'a, str>` because of the iteration but it's still an excellent improvement (and it's called every time)! 🥇